build.py

```diff
@@ -341,6 +344,91 @@ def step_end():
         subprocess.run((cargo_args + ["--", "--qsharp-json", project]), check=True, text=True, cwd=root_dir)
     step_end()
 
+if build_notebooks:
+    step_start("Running notebook samples")
+    # Find all notebooks in the samples directory. Skip the basic sample and the azure submission sample, since those won't run
+    # nicely in automation.
+    notebook_files = [os.path.join(dp, f) for dp, _, filenames in os.walk(samples_src) for f in filenames if f.endswith(".ipynb")
+                      and not (f.startswith("sample.") or f.startswith("azure_submission."))]
+    # Check if in a virtual environment
+    if (
+        os.environ.get("VIRTUAL_ENV") is None
+        and os.environ.get("CONDA_PREFIX") is None
+        and os.environ.get("CI") is None
+    ):
+        print("Not in a virtual python environment")
+
+        venv_dir = os.path.join(samples_src, ".venv")
+        # Create virtual environment under repo root
+        if not os.path.exists(venv_dir):
+            print(f"Creating a virtual environment under {venv_dir}")
+            venv.main([venv_dir])
+
+        # Check if .venv/bin/python exists, otherwise use .venv/Scripts/python.exe (Windows)
+        python_bin = os.path.join(venv_dir, "bin", "python")
+        if not os.path.exists(python_bin):
+            python_bin = os.path.join(venv_dir, "Scripts", "python.exe")
+        print(f"Using python from {python_bin}")
+    else:
+        # Already in a virtual environment, use current Python
+        python_bin = sys.executable
+
+    # copy the process env vars
+    pip_env: dict[str, str] = os.environ.copy()
+
+    # Install the qsharp package
+    pip_install_args = [
+        python_bin,
+        "-m",
+        "pip",
+        "install",
+        "-e",
```

Comment:
This will rebuild the Rust module from scratch. Also the install doesn't need to be editable. Elsewhere in the file (when we run the pip package tests) we install the package from the already built wheel. I'd recommend doing that instead. (Same for the widgets package).

Reply:
It actually doesn't rebuild it from scratch, but rather benefits from the incremental builds. So in practice the Rust compilation is nearly zero if the binary was built before and then it just links to the Python files in this folder rather than needing to zip everything into a wheel and then unzip again into the venv's module folder. In the ubuntu build pipeline this took about 2 seconds since it didn't need to rebuild anything. And if run in isolation (without the `--pip` flag) this line will build the needed Rust components if they haven't been recently built. I prefer that to requiring that folks remember to specify both `--pip` and the command runs the samples.

Reply:
There's some demonic behavior that occurs with `pip install` where it will corrupt the `.whl` file that we built in the previous step. I had to fix this in #539 and I strongly suggest you avoid it as well :)

Reply:
oh, good tip! I'll avoid this pattern then.

Reply:
From local testing, it looks like a non-editable `pip install` with the folder path to widgets src does what we want (I checked the targets/wheel folder and saw the wheel built earlier is not touched when using this approach).